core/kt-osrd-path/src/main/kotlin/fr/sncf/osrd/path/implementations/TrainPathBuilder.kt

```diff
@@ -0,0 +1,325 @@
+    for ((i, value) in merged.withIndex()) {
+        // In normal cases, we always include lower bound and never include upper bound.
+        // Exceptions: the very first and last bounds are included, and a zero-length range is
+        // included in both directions. The next range then needs to have its lower bound excluded.
+        // Two consecutive zero-length ranges are not supported.
+        val includeLowerEndpoint = i == 0 || !zeroLengthIndexes[i - 1]
+        val includeUpperEndpoint = i == merged.size - 1 || zeroLengthIndexes[i]
+        fun mapBoundType(included: Boolean) = if (included) BoundType.CLOSED else BoundType.OPEN
+        val range =
+            Range.range(
+                prevRangeLength,
+                mapBoundType(includeLowerEndpoint),
+                prevRangeLength + value.length,
+                mapBoundType(includeUpperEndpoint),
+            )
+        builder.put(range, value)
+        prevRangeLength += value.length
+    }
+    return builder.build()
+}
+
+/** Generate the chunk ranges from given block ranges. */
+private fun generateTrackChunks(
+    rawInfra: RawInfra,
+    blockInfra: BlockInfra,
+    blocks: LinearObjectMap<BlockRange>,
+): LinearObjectMap<DirChunkRange> {
+    val res = mutableListOf<DirChunkRange>()
+    for (blockRange in blocks.asMapOfRanges().values) {
+        var currentChunkOffset = Offset<Block>(0.meters)
+        for (chunk in blockInfra.getTrackChunksFromBlock(blockRange.value)) {
+            val chunkLength = rawInfra.getTrackChunkLength(chunk.value)
+
+            val chunkStart = Offset<TrackChunk>(blockRange.from - currentChunkOffset)
+            val chunkEnd = Offset<TrackChunk>(blockRange.to - currentChunkOffset)
+            if (
+                chunkStart <= chunkEnd &&
+                    Offset.zero<TrackChunk>() <= chunkEnd &&
+                    chunkStart <= chunkLength
+            ) {
+                val chunkRange =
+                    DirChunkRange(chunk, max(chunkStart, Offset.zero()), min(chunkEnd, chunkLength))
+                res.add(chunkRange)
+            }
+
+            currentChunkOffset += chunkLength.distance
+        }
+    }
+
+    // We need to filter out zero-length ranges that aren't first or last
+    val filtered = mutableListOf<DirChunkRange>()
+    for ((i, chunkRange) in res.withIndex()) {
+        if (i == 0 || i == res.size - 1 || chunkRange.length > 0.meters) filtered.add(chunkRange)
+    }
+    return buildRangeMap(filtered)
+}
+
+/**
+ * Generate the route ranges from given chunk ranges, with actual route IDs given as input. This
+ * just maps the offsets and precise ranges.
+ */
+private fun generateRouteRanges(
+    rawInfra: RawInfra,
+    chunks: LinearObjectMap<DirChunkRange>,
+    routes: List<RouteId>,
+): LinearObjectMap<RouteRange> {
+    val res = mutableListOf<RouteRange>()
+    val mappedChunks = chunks.asMapOfRanges().map { it.value }.associateBy { it.value }
```

Comment:
Cant' we just use a set? Something along
```suggestion
    val chunkSet = chunks.asMapOfRanges().map { it.value }.toSet()
```

Reply:
It's a `Map<DirTrackChunk, DirChunkRange>`. We need it to know which range of each chunk is actually used.